modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityStressUpdateBase.C

```diff
@@ -113,86 +126,128 @@ CrystalPlasticityStressUpdateBase::getSlipSystems()
+               "not consistent with expectations for "
+               "HCP crystal hexagonal lattices.");
+  else if (reader.getData(0).size() != miller_bravis_indices * 2)
+    mooseError("CrystalPlasticityStressUpdateBase Error: The number of entries in the first row of "
+               "the slip system file is not consistent with the expectations for the 4-index "
+               "Miller-Bravis assumption for HCP crystals. This file should represent both the "
+               "slip plane normal and the slip direction with 4-indices each.");
+
+  // set up the tranformation matrices
+  RankTwoTensor transform_planes, transform_directions;
+  transform_planes(0, 0) = 1.0 / _unit_cell_dimension[0];
+  transform_planes(1, 0) = 1.0 / (_unit_cell_dimension[0] * std::sqrt(3.0));
+  transform_planes(1, 1) = 2.0 / (_unit_cell_dimension[0] * std::sqrt(3.0));
+  transform_planes(2, 2) = 1.0 / (_unit_cell_dimension[2]);
+
+  transform_directions(0, 0) = _unit_cell_dimension[0];
+  transform_directions(0, 1) = -1.0 / (2 * _unit_cell_dimension[0]);
+  transform_directions(1, 1) = std::sqrt(3.0) * _unit_cell_dimension[0] / 2.0;
+  transform_directions(2, 2) = _unit_cell_dimension[2];
 
-  MooseUtils::checkFileReadable(vector_file_name);
-
-  fileslipsys.open(vector_file_name.c_str());
-
-  for (unsigned int i = 0; i < number_dislocation_systems; ++i)
+  for (unsigned int i = 0; i < _number_slip_systems; ++i)
   {
-    // Read the plane normal
-    for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
-      if (!(fileslipsys >> vec[j]))
-        mooseError(
-            "CrystalPlasticityStressUpdateBase Error: Premature end of file reading plane normal "
-            "vectors from the file ",
-            vector_file_name);
-
-    // Normalize the vectors
-    Real magnitude = 0.0;
-    for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
-      magnitude += Utility::pow<2>(vec[j]);
-    magnitude = std::sqrt(magnitude);
-
-    for (unsigned j = 0; j < LIBMESH_DIM; ++j)
-      plane_normal_vector(i * LIBMESH_DIM + j) = vec[j] / magnitude;
-
-    // Read the direction
-    for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
-      if (!(fileslipsys >> vec[j]))
-        mooseError("CrystalPlasticityStressUpdateBase Error: Premature end of file reading "
-                   "direction vectors "
-                   "from the file ",
-                   vector_file_name);
-
-    // Normalize the vectors
-    magnitude = Utility::pow<2>(vec[0]) + Utility::pow<2>(vec[1]) + Utility::pow<2>(vec[2]);
-    magnitude = std::sqrt(magnitude);
+    // read in raw data from file and store in the temporary vectors
+    for (unsigned int j = 0; j < reader.getData(i).size(); ++j)
+    {
+      if (j < miller_bravis_indices)
+      {
+        // Planes are directly copied over, per a_1 = convention used here
+        if (j < 2)
+          temporary_slip_plane[j] = reader.getData(i)[j];
+        else if (j == 3)
+          temporary_slip_plane[j-1] = reader.getData(i)[j];
```

Comment:
I am guessing the `reader.getData(i)[2]` depends on the indices from `reader.getData(i)[0]` and `reader.getData(i)[1]`. It maybe helpful to add a comment here to make it clear and maybe check the data in `reader.getData(i)[2]` too to make sure that we have valid input?

Reply:
Both good points. I will do both--thanks!